src/chronify/datetime_range_generator.py

```diff
@@ -65,8 +70,95 @@ def iter_timestamps(self) -> Generator[datetime, None, None]:
                     ):
                         yield pd.Timestamp(cur)
 
-    def list_distinct_timestamps_from_dataframe(self, df: pd.DataFrame) -> list[datetime]:
-        return sorted(df[self._model.time_column].unique())
-
     def list_time_columns(self) -> list[str]:
         return self._model.list_time_columns()
+
+    def list_distinct_timestamps_from_dataframe(self, df: pd.DataFrame) -> list[datetime]:  # TODO
+        result = sorted(df[self._model.time_column].unique())
+        if not isinstance(result[0], datetime):
+            result = [pd.Timestamp(x) for x in result]
+        return result
+
+
+class DatetimeRangeGenerator(DatetimeRangeGeneratorBase):
+    """Generates datetime ranges based on a DatetimeRange model."""
+
+    def __init__(
+        self,
+        model: DatetimeRange,
+        leap_day_adjustment: Optional[LeapDayAdjustmentType] = None,
+    ) -> None:
+        super().__init__(model, leap_day_adjustment=leap_day_adjustment)
+        assert isinstance(self._model, DatetimeRange)
+
+    def list_timestamps(self) -> list[datetime]:
+        return list(self._iter_timestamps())
+
+
+class DatetimeRangeGeneratorExternalTimeZone(DatetimeRangeGeneratorBase):
+    """Generates datetime ranges based on a DatetimeRangeWithTZColumn model.
+    datetime ranges will be tz-naive and can be listed by time_zone name using special class func
+    These ranges may be localized by the time_zone name.
+    # TODO: add offset as a column
+    """
+
+    def __init__(
+        self,
+        model: DatetimeRangeWithTZColumn,
+        leap_day_adjustment: Optional[LeapDayAdjustmentType] = None,
+    ) -> None:
+        super().__init__(model, leap_day_adjustment=leap_day_adjustment)
+        assert isinstance(self._model, DatetimeRangeWithTZColumn)
+        if self._model.get_time_zones() == []:
+            msg = (
+                f"DatetimeRangeWithTZColumn.time_zones needs to be instantiated for "
+                f"DatetimeRangeGeneratorExternalTimeZone: {self._model}"
+            )
+            raise InvalidValue(msg)
+
+    def _list_timestamps(self, time_zone: Optional[tzinfo]) -> list[datetime]:
+        """always return tz-naive timestamps relative to input time_zone"""
+        if self._model.start_time_is_tz_naive():
+            if time_zone:
+                start = self._model.start.replace(tzinfo=time_zone)
+            else:
+                start = None
+        else:
+            if time_zone:
+                start = self._model.start.astimezone(time_zone)
+            else:
+                start = self._model.start.replace(tzinfo=None)
+        timestamps = list(self._iter_timestamps(start=start))
+        return [x.replace(tzinfo=None) for x in timestamps]
+
+    def list_timestamps(self) -> list[datetime]:
```

Comment:
I am getting confused on when duplicates are allowed. For consideration, should this be called list_distinct_timestamps?

Reply:
Changing to a flattened version of dct.values() from list_timestamps_by_time_zone.

Reply:
I'm still struggling to see the use case when there are multiple time zones. What will the caller do with them? They will always prefer list_timestamps_by_time_zone.